File: Assignment 4/C/06. ll left-shift by k nodes.c
// 2->4->7->8->9
// K=3
// 8->9->2->4->7

#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    int val;
    struct node* next;
}node;

node* createNode(int n){
    node* t = (node*)malloc(sizeof(node));
    t->val = n;
    t->next = NULL;
    return t;
}

void push(node* head,int n){
    node* t = (node*)malloc(sizeof(node));
    if(t==NULL){
        printf("no space for %d",n);
        return;
    }
    t->val = n;
    t->next = NULL;
    while(head->next){
        head = head->next;
    }
    head->next = t;
}

node* leftShift(node* head, int k){
    node* last = head;
    while(last->next){
        last = last->next;
    }
    last->next = head;

    node* curr = head;
    node* prev = head;
    for(int i=0; i<k; i++){
        prev = curr;
        curr = curr->next;
    }
    prev->next = NULL;
    return curr;
}

void traversal(node* head){
    node* ptr = head;
    while(ptr){
        printf("%d --> ",ptr->val);
        ptr = ptr->next;
    }
    printf("NULL\n");
}

int main(){
    node* head = createNode(2);
    push(head,4);
    push(head,7);
    push(head,8);
    push(head,9);
    traversal(head);

    int k;
    printf("enter k: ");
    scanf("%d",&k);

    head = leftShift(head,k);
    traversal(head);
    
    return 0;
}


//OUTPUT:

// 2 --> 4 --> 7 --> 8 --> 9 --> NULL

File: Assignment 4/Python/06. ll left-shift by k nodes .py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def create_node(n):
    return Node(n)

def push(head, n):
    new_node = Node(n)
    if not head:
        return new_node
    ptr = head
    while ptr.next:
        ptr = ptr.next
    ptr.next = new_node
    return head

def left_shift(head, k):
    if head is None or head.next is None or k <= 0:
        return head

    length = 1
    last = head
    while last.next:
        last = last.next
        length += 1

    last.next = head

    k = k % length
    new_last = head
    for _ in range(length - k - 1):
        new_last = new_last.next

    new_head = new_last.next
    new_last.next = None

    return new_head

def traversal(head):
    ptr = head
    while ptr:
        print(f"{ptr.val} --> ", end="")
        ptr = ptr.next
    print("NULL")

if __name__ == "__main__":
    head = create_node(2)
    head = push(head, 4)
    head = push(head, 7)
    head = push(head, 8)
    head = push(head, 9)

    print("Original list:")
    traversal(head)

    k = int(input("Enter k: "))
    head = left_shift(head, k)

    print("List after left shift:")
    traversal(head)

# OUTPUT:
# Original list:
# 2 --> 4 --> 7 --> 8 --> 9 --> NULL
# Enter k: 3
# List after left shift:
# 7 --> 8 --> 9 --> 2 --> 4 --> NULL
